### ISSUE TITLE
iter(...) is not revealed the same as __iter__ method

### ISSUE THREAD

**Issue by loicdiridollou** [astral-sh/ty#3041]
### Summary

Hi team!

We are facing a small issue with `ty` on the `pandas-stubs` repo that struggles to match the identified type when using `__iter__` method vs the `iter(...)` calls.
We have made a small playground for it https://play.ty.dev/25fa719b-09c8-4c3d-bc16-92deb1279c03.
Not sure exactly what is going on in the backend, curious to hear your insights because from intuition `iter(...)` and `__iter__` should match.

Please let us know!

```python
from typing import Any, Generic, Never, TypeVar, reveal_type

_T_EXPANDING = TypeVar("_T_EXPANDING")
_T_STR = TypeVar("_T_STR", bound=Series)


class StringMethods(Generic[_T_EXPANDING, _T_STR]):
    def __getitem__(self, key: int) -> _T_STR: ...
    def __iter__(self) -> Never: ...

class Series:
    @property
    def str(self) -> StringMethods[Any, Series]: ...


def _0() -> None:
    reveal_type(StringMethods().__iter__())  # all 4 type checkers give Never

def _1() -> None:
    reveal_type(iter(Series().str))  # mypy: Any, pyright, pyrefly: Never, ty: Unknown
```
FWIW mypy also struggles with it while `pyright` and `pyrefly` correctly match the types.

### Version

0.0.23

**Comment by carljm** [comment]
Thanks for the report! I think perhaps some other factor is confounding your mypy test; when I simplify to this reproduction, ty is alone in inferring `Unknown`; all other type checkers infer `Never`:

```py
from typing import reveal_type, Never

class StringMethods:
    def __iter__(self) -> Never:
        raise RuntimeError()

reveal_type(iter(StringMethods()))
```

@dcreager it looks like the root cause of this issue is that we end up with a constraint set where the only constraint is `Never <= T`, and we explicitly treat a lower bound of `Never` as no bound at all. So we end up with "no constraints", and rather than solving to `Never` we fall back to `Unknown`.

It is true that technically any solution for `T` here is not-wrong, but we shouldn't treat this case the same as a typevar we found no information about -- we should prefer the narrowest possible solution, which is `Never`. IOW we have to treat "no constraints found" differently from "constraint found which doesn't actually constrain the type at all."

Here's a standalone repro of the same issue, without involving stdlib types or functions:

```py
from typing import Never, Protocol, reveal_type

class ReturnsT[T](Protocol):
    def get(self) -> T: ...

class ReturnsNever:
    def get(self) -> Never:
        raise RuntimeError

def through_protocol[T](x: ReturnsT[T]) -> T:
    return x.get()

reveal_type(through_protocol(ReturnsNever()))
```
